mqtt/mqtt-broker/src/broker.rs

```diff
@@ -100,6 +104,62 @@ where
         Ok(self.snapshot())
     }
 
+    fn prepare_activities(client_id: &ClientId, session: &Session) -> Vec<(ClientId, Activity)> {
+        let mut activities = Vec::new();
+        for sub in session
+            .subscriptions()
+            .into_iter()
+            .flat_map(HashMap::values)
+        {
+            if let Ok(client_info) = session.client_info() {
+                let sub_topic_filter = sub.filter().to_string();
+                let operation = Operation::new_subscribe(proto::SubscribeTo {
+                    topic_filter: sub_topic_filter.clone(),
+                    qos: *sub.max_qos(),
+                });
+                activities.push((
+                    client_id.clone(),
+                    Activity::new(client_id.clone(), client_info.clone(), operation),
+                ));
+            }
+        }
+        activities
+    }
+
+    fn reevaluate_subscriptions(&mut self) -> Result<(), Error> {
+        let disconnecting: Vec<ClientId> = self
+            .sessions
+            .iter()
+            .flat_map(|(client_id, session)| Self::prepare_activities(client_id, session))
+            .filter_map(
+                |(client_id, activity)| match self.authorizer.authorize(activity) {
+                    Ok(Authorization::Allowed) => None,
+                    Ok(Authorization::Forbidden(reason)) => {
+                        debug!(
+                            "client {} not allowed to subscribe to topic. {}",
+                            client_id, reason
+                        );
+                        Some(client_id)
+                    }
+                    Err(e) => {
+                        warn!(message="error authorizing client subscription: {}", error = %e);
+                        Some(client_id)
+                    }
+                },
+            )
+            .collect();
```

Comment:
actually you don't need to call collect here. just iterate on the filter_map result